src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -155,12 +162,17 @@ double LMNNFunction<MetricType>::Evaluate(const arma::mat& coordinates,
       // Bound constraints to avoid uneccesary computation.
       for (size_t l = 0, bp = k; l < bp ; l++)
       {
-        // Calculate cost due to data point, target neighbors, impostors
+        // Check if we already have distance between impostor & data point
+        // stored.
+        double distImp = distance(l, i) > 0 ? distance(l, i) :
```

Comment:
You could change the condition `distance(l, i) > 0` to `iteration % range == 0`, and then avoid the `arma::fill::zeros` in the constructor of the `distance` matrix.  Plus then we don't have to worry about the strange edge case when `distance(l, i)` is actually legitimately 0. :)

Reply:
Had exact same thoughts, but then proceeded with this one :) 
Will change it shortly. I also timed this approach and it look on iris we get a speed up of .1 sec

Reply:
Right, I think in many cases this could be a major speedup.  But I don't see any way to avoid the similar `targetNeighbors` distance computations.

Reply:
Actually are you sure the speedup is 0.1s?  In the table you showed iris only takes 0.02s.  If we made the computation take negative time I think we are definitely on to something good ;)

Reply:
Now I have to admit my system is no more than turtle, the second timing is on my system. And most probaby some other parameters :)

Reply:
Ah, ok, different systems, no worries :)